smart-contract-feature-development.md

```diff
@@ -0,0 +1,94 @@
+# Purpose
+
+The purpose of this document is to provide guidance on how develop smart contracts without the need for feature branches.
+
+# Summary
+
+Feature branches result in a lot of effort spent keeping the feature branch up to date. We want to have modular code
+that enables us to easily compose features together such that we can get code mainlined early and toggle it off when
+running in production.
+
+# Problem Statement + Context
+
+We want to move away from a world with long lived feature branches.
+To do so, we need to have some flexibility when it comes to composition of features.
+This is easy in off-chain code as the cost is minimal for toggling off certain features but still having them in the source code.
+This is not the case for Solidity. This document exists to outline how to think about composition of features for Solidity code.
+When the feature is ready to be shipped, the feature specific contract should be merged into the base contract.
+
+# Alternatives Considered
+
+## Copy and Paste Implementation
+
+This was the solution for the fault proofs project, which worked well until the custom gas token feature was introduced. Now the contracts
+are in a state such that the custom gas token feature needs to be ported to the `OptimismPortal2` before the `OptimismPortal` and the
+`OptimismPortal2` can be merged. This may be a good solution for security critical code as we do not rely on the correctness
+of solidity inheritance and all of the logic lives in a single contract, making it easier to audit.
+
+# Proposed Solution
+
+Below includes general guidelines for how to develop solidity feature contracts that are meant to be merged into mainline at
+some point in the future.
+
+### Composition through Inheritance
```

Comment:
have we considered composition through solidity libraries? I haven't thought deeply about this but given that they are stateless, they may be easier to reason about, and reduce the complexity of storage concerns.

Reply:
This is actually how the custom gas token feature was implemented, see [here](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/libraries/GasPayingToken.sol). That pattern made it easy to share functionality between multiple contracts and only write the storage related code once. I do like this pattern and there are likely opportunities to use it in the future. It isn't immediately obvious if we want to enshrine this pattern everywhere, it would become part of the style guide to make storage happen through libraries that expose getters to storage

Reply:
Cool. I think we agree it's a useful tool for certain jobs.